test/integration/test_cluster.c

```diff
@@ -170,3 +170,31 @@ TEST(cluster, dataOnNewNode, setUp, tearDown, 0, cluster_params)
 	clientCloseRows(&rows);
 	return MUNIT_OK;
 }
+
+TEST(cluster, hugeRow, setUp, tearDown, 0, NULL)
+{
+	struct fixture *f = data;
+	unsigned stmt_id;
+	unsigned last_insert_id;
+	unsigned rows_affected;
+	char *sql;
+	ssize_t n;
+size_t huge = 20000000;
+	(void)params;
+
+	HANDSHAKE;
+	OPEN;
+	PREPARE("CREATE TABLE IF NOT EXISTS model(key TEXT, value TEXT, UNIQUE(key))", &stmt_id);
+	EXEC(stmt_id, &last_insert_id, &rows_affected);
+	sql = munit_malloc(huge);
+	n = snprintf(sql, huge, "INSERT OR REPLACE INTO model (key, value) VALUES('my-key-1', '");
+	memset(sql + n, 'A', huge - n);
+	memcpy(sql + huge - 3, "')", 3);
+	PREPARE(sql, &stmt_id);
+	free(sql);
+	EXEC(stmt_id, &last_insert_id, &rows_affected);
+	/* Again */
+	EXEC(stmt_id, &last_insert_id, &rows_affected);
+	return MUNIT_OK;
```

Comment:
I'm a bit surprised that the there's no test failure or assertion violation because of the prepared statement above not being finalized.

But probably we have changed that behavior and now we automatically finalize leftover statements.

Reply:
They're finalized when the statement registry is closed here https://github.com/canonical/dqlite/blob/ee7a3cdb3a996fa798443b9df51391925a12b8d8/src/gateway.c#L72